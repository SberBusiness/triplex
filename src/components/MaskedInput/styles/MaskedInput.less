@import 'src/styles/style';

.maskedInputWrapper {
    position: relative;

    .maskedInputPlaceholder {
        position: absolute;
        left: 0;
        top: 0;

        &::placeholder {
            // Переопределяет opacity: 0 в FormFieldInput.less.
            opacity: 1;
        }
    }
}

.maskedInput {
    display: block;
    position: relative;
    background: none !important;
}



.grouped {
    flex: 1;

    &:not(:first-child) {
        margin-left: -1px;
    }

    .left {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
    }

    .intermediate {
        border-radius: 0;
    }

    .right {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
    }

    .maskedInput.error {
        z-index: 1;
    }

    .maskedInput:focus {
        z-index: 2;
    }
}
