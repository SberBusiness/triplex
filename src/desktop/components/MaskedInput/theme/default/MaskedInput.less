@import '../../../../common/theme/default';

.maskedInputWrapper {
    position: relative;

    &:disabled {
        .inputDisabled();
    }
}

.maskedInput {
    position: relative;
    background: none !important;
}

.maskedInputPlaceholder {
    color: @color-asphalt-01 !important;
    border-color: transparent !important;
    box-shadow: none !important;
    position: absolute;
    left: 0;
    top: 0;
}

.grouped {
    flex: 1;

    &:not(:first-child) {
        margin-left: -1px;
    }

    .left {
        border-top-right-radius: 0;
        border-bottom-right-radius: 0;
    }

    .intermediate {
        border-radius: 0;
    }

    .right {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
    }

    .maskedInput.error {
        z-index: 1;
    }

    .maskedInput:focus {
        z-index: 2;
    }
}
