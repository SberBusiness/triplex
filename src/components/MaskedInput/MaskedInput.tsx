import React, {useState, useEffect, useCallback} from 'react';
import MaskedInputTextMask, {conformToMask, MaskedInputProps} from 'react-text-mask';
import {Input} from '../Input/Input';
import {EInputGroupPosition} from '../InputGroup/InputGroup';
import {classnames} from '@sberbusiness/triplex/utils/classnames/classnames';
import {presets, TMaskedInputPresets} from './presets';

/** Маска. Каждый элемент массива должен быть либо строкой, либо регулярным выражением. Каждая строка — это фиксированный символ в маске, а каждое регулярное выражение — это заполнитель, который принимает пользовательский ввод.
 * Подробнее можно ознакомиться https://github.com/text-mask/text-mask/blob/master/componentDocumentation.md#mask.
 * */
export type TMaskedInputMask = Array<string | RegExp>;

export interface IMaskedInputProps extends Omit<MaskedInputProps, 'guide' | 'mask' | 'render'> {
    value: string;
    /** Состояние ошибки. */
    error?: boolean;
    /** Позиция внутри компонента InputGroup. */
    groupPosition?: EInputGroupPosition;
    forwardRef?: React.MutableRefObject<HTMLElement | null>;
    /** Маска. Каждый элемент массива должен быть либо строкой, либо регулярным выражением. Каждая строка — это фиксированный символ в маске, а каждое регулярное выражение — это заполнитель, который принимает пользовательский ввод.
     * Подробнее можно ознакомиться https://github.com/text-mask/text-mask/blob/master/componentDocumentation.md#mask.
     * */
    mask: TMaskedInputMask;
    /** Плейсхолдер, отображаемый при вводе. Например: дд.мм.гггг, при вводе будет отображаться как 22.1м.гггг. */
    placeholderMask?: string;
}

export interface IIMaskedInputFC extends React.FC<IMaskedInputProps> {
    presets: TMaskedInputPresets;
}

const mapInputGroupPositionToCSSClass = {
    [EInputGroupPosition.LEFT]: 'cssClass[left]',
    [EInputGroupPosition.INTERMEDIATE]: 'cssClass[intermediate]',
    [EInputGroupPosition.RIGHT]: 'cssClass[right]',
};

/**
 * Компонент маскированного ввода.
 * Основан на https://github.com/text-mask/text-mask.
 */
export const MaskedInput: IIMaskedInputFC = ({
    className,
    forwardRef,
    disabled,
    groupPosition,
    mask,
    onChange,
    placeholder,
    placeholderChar = '0',
    placeholderMask,
    value,
    style,
    ...inputProps
}) => {
    // Значение инпута, отображающего часть введенного значения и оставшуюся маску.
    const [placeholderValue, setPlaceholderValue] = useState('');

    useEffect(() => {
        /**
         * Возвращает значение placeholderValue.
         */
        const calculatePlaceholderValue = (): string => {
            // Массив символов placeholderValue.
            let nextPlaceholderValue: string[] = [];

            // Значение инпута остутствует.
            if (!value) {
                // Передан props placeholderMask, например дд.мм.гггг
                if (placeholderMask) {
                    // При наличии маски плейсхолдера, placeholderValue равен маски плейсхолдера.
                    nextPlaceholderValue = placeholderMask.split('');
                } else {
                    // Маска с символами заполнения, например 00.00.00
                    const {conformedValue} = conformToMask('', mask, {guide: true, placeholderChar});
                    nextPlaceholderValue = conformedValue.split('');
                }
            } else {
                // Инпут имеет value.

                // Value с маской, например: 22.00.00
                const {conformedValue} = conformToMask(value.toString(), mask, {guide: true, placeholderChar});

                // Символы placeholderValue собираются из введенного пользователем значения и оставшейся части из placeholderMask или placeholderChar.
                for (let i = 0; i < mask.length; i++) {
                    // Не редактируемый символ маски.
                    if (typeof mask[i] === 'string') {
                        // Символ из маски.
                        nextPlaceholderValue[i] = conformedValue[i];
                    } else {
                        // Не введенный пользователем символ заполняется символом placeholderMask или placeholderChar.
                        if (conformedValue[i] === placeholderChar && !value.toString()[i]) {
                            nextPlaceholderValue[i] = placeholderMask?.[i] || placeholderChar;
                        } else {
                            nextPlaceholderValue[i] = conformedValue[i];
                        }
                    }
                }
            }

            return nextPlaceholderValue.join('');
        };

        setPlaceholderValue(calculatePlaceholderValue());
    }, [value, mask, placeholderChar, placeholderMask]);

    const handleChange = useCallback(
        (event: React.ChangeEvent<HTMLInputElement>) => {
            const {value: nextValue} = event.target;

            if (value !== nextValue) {
                onChange?.(event);
            }
        },
        [value, onChange]
    );

    // Возвращает value в случае, если маска является номером телефона.
    const getPhoneValue = (): string => {
        // Выражение, для поиска чисел вида {любая цифра}7, например 87, 971 и т.д.
        const regEx = /^\d7/;
        if (regEx.test(value)) {
            // Если вторая цифра номера 7, добавляется +7 перед этим, иначе conformToMask вместо 9701234567 вернет +7901234567.
            value = '+7' + value;
        }

        return conformToMask(value, mask, {guide: false, placeholderChar}).conformedValue;
    };

    // Возвращает value, для передачи в компоненты рендера. Для некоторых типов масок, value приходится модифицировать из-за багов.
    const getValue = (): string => {
        if (mask === MaskedInput.presets.masks.phone) {
            return getPhoneValue();
        }

        return conformToMask(value, mask, {guide: false, placeholderChar}).conformedValue;
    };

    return (
        <div className={classnames('cssClass[maskedInputWrapper]', {'cssClass[grouped]': !!groupPosition})}>
            {/* Input, отображающий маску. */}
            <Input
                className={classnames(
                    'cssClass[maskedInputPlaceholder]',
                    groupPosition && mapInputGroupPositionToCSSClass[groupPosition],
                    className
                )}
                disabled={disabled}
                onChange={() => {}}
                placeholder={!value && placeholder ? '' : placeholderValue}
                readOnly
                aria-hidden="true"
                tabIndex={-1}
                style={style}
            />

            {/* Input, отображающий введенное значение. */}
            <MaskedInputTextMask
                className={classnames(
                    'cssClass[maskedInput]',
                    groupPosition && mapInputGroupPositionToCSSClass[groupPosition],
                    {'cssClass[error]': Boolean(inputProps.error)},
                    className
                )}
                // https://github.com/text-mask/text-mask/pull/993
                defaultValue=""
                disabled={disabled}
                /* Input отображает только введенное значение без маски, маска рисуется в inputPlaceholder. */
                guide={false}
                render={(ref, props) => (
                    <Input
                        ref={(el) => {
                            if (el) {
                                ref(el);
                                forwardRef && (forwardRef.current = el);
                            }
                        }}
                        {...props}
                        placeholder={placeholder || ''}
                    />
                )}
                mask={mask}
                onChange={handleChange}
                placeholderChar={placeholderChar}
                // value={value} не используется т.к. возникает баг при передаче снаружи изначально пустого value, а затем не пустого.
                value={getValue()}
                style={style}
                {...inputProps}
            />
        </div>
    );
};

MaskedInput.presets = presets;
