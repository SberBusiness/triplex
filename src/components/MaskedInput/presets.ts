import {TMaskedInputMask} from './MaskedInput';

export interface TMaskedInputPresets {
    // Предзаготовленный набор плейсхолдеров масок.
    placeholderMasks: {
        // Номер автомобиля A000AA 00.
        carNumber: string;
        // Номер карты 0000 0000 0000 0000 (от 16 до 20 символов).
        cardNumber: string;
        // Дата дд.мм.гггг.
        date: string;
        // ИНН 0000000000 (от 10 до 12 символов).
        inn: string;
        // Доб. номер 000 (от 3 до 4 символов).
        phoneExtension: string;
        // УИН 00000000000000000000 (от 20 до 25 символов).
        uin: string;
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: string;
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: string;
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: string;
    };
    // Предзаготовленный набор масок.
    masks: {
        // Номер счета 00000 000 0 00000000000.
        account: TMaskedInputMask;
        // БИК 000 000 000.
        bic: TMaskedInputMask;
        // Номер автомобиля A000AA 00.
        carNumber: TMaskedInputMask;
        // Номер карты 0000 0000 0000 0000 (от 16 до 20 символов).
        cardNumber: TMaskedInputMask;
        // Дата.
        date: TMaskedInputMask;
        // Водительское удостоверение 00 00 000000.
        driversLicense: TMaskedInputMask;
        // ИНН 0000000000 (от 10 до 12 символов).
        inn: TMaskedInputMask;
        // КБК 00000000000000000000.
        kbk: TMaskedInputMask;
        // КПП 000000000.
        kpp: TMaskedInputMask;
        // Широта 00.000000.
        latitude: TMaskedInputMask;
        // Долгота 00.000000.
        longitude: TMaskedInputMask;
        // ОГРН 0000000000000.
        ogrn: TMaskedInputMask;
        // ОКТМО 00000000.
        oktmo: TMaskedInputMask;
        // Паспорт РФ.
        passport: {
            // Код подразделения 000-000.
            departmentCode: TMaskedInputMask;
            // Номер 000000.
            number: TMaskedInputMask;
            // Серия 00 00.
            series: TMaskedInputMask;
        };
        // Номер телефона.
        phone: TMaskedInputMask;
        // Доб. номер 000 (от 3 до 4 символов).
        phoneExtension: TMaskedInputMask;
        // Индекс 000000.
        postalCode: TMaskedInputMask;
        // СНЛС 000-000-000 00.
        snils: TMaskedInputMask;
        // УИН 00000000000000000000 (от 20 до 25 символов).
        uin: TMaskedInputMask;
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: TMaskedInputMask;
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: TMaskedInputMask;
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: TMaskedInputMask;
    };
}

export const presets: TMaskedInputPresets = {
    // Предзаготовленный набор масок.
    masks: {
        account: [
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
        ],
        // БИК 000000000.
        bic: [/\d/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // Номер автомобиля A000AA 00.
        carNumber: [/[а-яА-ЯЁё]/, /\d/, /\d/, /\d/, /[а-яА-ЯЁё]/, /[а-яА-ЯЁё]/, /\d/, /\d/, /\d/],
        // Номер карты 0000 0000 0000 0000 0000.
        cardNumber: [
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
        ],
        date: [/\d/, /\d/, '.', /\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/],
        // Водительское удостоверение 00 00 000000.
        driversLicense: [/\d/, /\d/, ' ', /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // ИНН 0000000000 (от 10 до 12 символов).
        inn: Array<RegExp>(12).fill(/\d/),
        // КБК 00000000000000000000.
        kbk: Array<RegExp>(20).fill(/\d/),
        // КПП 000000000.
        kpp: Array<RegExp>(9).fill(/\d/),
        // Широта 00.000000.
        latitude: [/\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // Долгота 00.000000.
        longitude: [/\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // ОГРН 0000000000000.
        ogrn: Array<RegExp>(13).fill(/\d/),
        // ОКТМО 00000000.
        oktmo: Array<RegExp>(8).fill(/\d/),
        // Паспорт РФ.
        passport: {
            // Код подразделения 000-000.
            departmentCode: [/\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/],
            // Номер 000000.
            number: Array<RegExp>(6).fill(/\d/),
            // Серия 00 00.
            series: [/\d/, /\d/, ' ', /\d/, /\d/],
        },
        // Номер телефона.
        phone: ['+', '7', ' ', '(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, '-', /\d/, /\d/],
        // Доб. номер 000 (от 3 до 4 символов).
        phoneExtension: Array<RegExp>(4).fill(/\d/),
        // Индекс 000000.
        postalCode: Array<RegExp>(6).fill(/\d/),
        // СНИЛС 000-000-000 00.
        snils: [/\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, ' ', /\d/, /\d/],
        // УИН 00000000000000000000 (от 20 до 25 символов).
        uin: Array<RegExp>(25).fill(/\d/),
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: [/\d/, /\d/, /[а-яА-ЯЁё]/, /[а-яА-ЯЁё]/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: [/\d/, /\d/, /[а-яА-ЯЁё]/, /[а-яА-ЯЁё]/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/, '-', /\d/, /\d/],
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: [
            /\d/,
            /\d/,
            /[а-яА-ЯЁё]/,
            /[а-яА-ЯЁё]/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            '-',
            /\d/,
            /\d/,
            '-',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
        ],
    },
    // Предзаготовленный набор плейсхолдеров масок. Пустые пробелы в конце - под дополнительные не подсвеченные символы.
    placeholderMasks: {
        // Номер автомобиля A000AA00.
        carNumber: 'A000AA00 ',
        // Номер карты 0000 0000 0000 0000 (от 16 до 20 символов).
        cardNumber: '0000 0000 0000 0000     ',
        // Дата дд.мм.гггг.
        date: 'дд.мм.гггг',
        // ИНН 0000000000 (от 10 до 12 символов).
        inn: '0000000000  ',
        // Доб. номер 000 (от 3 до 4 символов).
        phoneExtension: '000 ',
        // УИН 00000000000000000000 (от 20 до 25 символов).
        uin: '00000000000000000000     ',
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: '00АА000000',
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: '00АА000000-00',
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: '00АА000000-00-0000',
    },
};
