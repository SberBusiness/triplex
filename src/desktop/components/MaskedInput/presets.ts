import {TMaskedInputMask} from './MaskedInput';

export interface TMaskedInputPresets {
    // Предзаготовленный набор плейсхолдеров масок.
    placeholderMasks: {
        // Номер автомобиля A000AA 00.
        carNumber: string;
        // Номер карты 0000 0000 0000 0000 (от 16 до 20 символов).
        cardNumber: string;
        // Дата дд.мм.гггг.
        date: string;
        // ИНН 0000000000 (от 10 до 12 символов).
        inn: string;
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: string;
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: string;
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: string;
    };
    // Предзаготовленный набор масок.
    masks: {
        // Номер счета 00000 000 0 00000000000.
        account: TMaskedInputMask;
        // БИК 000 000 000.
        bic: TMaskedInputMask;
        // Номер автомобиля A000AA 00.
        carNumber: TMaskedInputMask;
        // Номер карты 0000 0000 0000 0000 (от 16 до 20 символов).
        cardNumber: TMaskedInputMask;
        // Дата.
        date: TMaskedInputMask;
        // Водительское удостоверение 00 00 000000.
        driversLicense: TMaskedInputMask;
        // ИНН 0000000000 (от 10 до 12 символов).
        inn: TMaskedInputMask;
        // КБК 00000000000000000000.
        kbk: TMaskedInputMask;
        // КПП 000000000.
        kpp: TMaskedInputMask;
        // Широта 00.000000.
        latitude: TMaskedInputMask;
        // Долгота 00.000000.
        longitude: TMaskedInputMask;
        // ОГРН 00000000000000000000.
        ogrn: TMaskedInputMask;
        // ОКТМО 00000000.
        oktmo: TMaskedInputMask;
        // Паспорт РФ.
        passport: {
            // Код подразделения 000-000.
            departmentCode: TMaskedInputMask;
            // Номер 000000.
            number: TMaskedInputMask;
            // Серия 00 00.
            series: TMaskedInputMask;
        };
        phone: TMaskedInputMask;
        // Индекс 000000.
        postalCode: TMaskedInputMask;
        // СНЛС 000-000-000 00.
        snils: TMaskedInputMask;
        // УИН 00000000000000000000.
        uin: TMaskedInputMask;
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: TMaskedInputMask;
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: TMaskedInputMask;
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: TMaskedInputMask;
    };
}

export const presets: TMaskedInputPresets = {
    // Предзаготовленный набор плейсхолдеров масок.
    placeholderMasks: {
        // Номер автомобиля A000AA00. Пустые пробелы в конце - под дополнительные не подсвеченные символы.
        carNumber: 'A000AA00 ',
        // Номер карты 0000 0000 0000 0000 (от 16 до 20 символов). Пустые пробелы в конце - под дополнительные не подсвеченные символы.
        cardNumber: '0000 0000 0000 0000     ',
        // Дата дд.мм.гггг.
        date: 'дд.мм.гггг',
        // ИНН 0000000000 (от 10 до 12 символов). Пустые пробелы в конце - под дополнительные не подсвеченные символы.
        inn: '0000000000  ',
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: '00АА000000',
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: '00АА000000-00',
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: '00АА000000-00-0000',
    },
    // Предзаготовленный набор масок.
    masks: {
        account: [
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
        ],
        // БИК 000000000.
        bic: [/\d/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // Номер автомобиля A000AA 00.
        carNumber: [/[а-яА-ЯЁё]/, /\d/, /\d/, /\d/, /[а-яА-ЯЁё]/, /[а-яА-ЯЁё]/, /\d/, /\d/, /\d/],
        // Номер карты 0000 0000 0000 0000 0000.
        cardNumber: [
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            ' ',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
        ],
        date: [/\d/, /\d/, '.', /\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/],
        // Водительское удостоверение 00 00 000000.
        driversLicense: [/\d/, /\d/, ' ', /\d/, /\d/, ' ', /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // ИНН 0000000000 (от 10 до 12 символов).
        inn: Array<RegExp>(12).fill(/\d/),
        // КБК 00000000000000000000.
        kbk: Array<RegExp>(20).fill(/\d/),
        // КПП 000000000.
        kpp: Array<RegExp>(9).fill(/\d/),
        // Широта 00.000000.
        latitude: [/\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // Долгота 00.000000.
        longitude: [/\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // ОГРН 00000000000000000000.
        ogrn: Array<RegExp>(20).fill(/\d/),
        // ОКТМО 00000000.
        oktmo: Array<RegExp>(8).fill(/\d/),
        passport: {
            // Код подразделения 000-000.
            departmentCode: [/\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/],
            // Номер 000000.
            number: Array<RegExp>(6).fill(/\d/),
            // Серия 00 00.
            series: [/\d/, /\d/, ' ', /\d/, /\d/],
        },
        phone: ['+', '7', ' ', '(', /[1-9]/, /\d/, /\d/, ')', ' ', /\d/, /\d/, /\d/, '-', /\d/, /\d/, '-', /\d/, /\d/],
        // Индекс 000000.
        postalCode: Array<RegExp>(6).fill(/\d/),
        // УИН 00000000000000000000.
        uin: Array<RegExp>(20).fill(/\d/),
        // СНИЛС 000-000-000 00.
        snils: [/\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/, ' ', /\d/, /\d/],
        // Единый лицевой счёт поставщика услуг (ЖКУ) 00АА000000.
        zhkuAccount: [/\d/, /\d/, /[а-яА-ЯЁё]/, /[а-яА-ЯЁё]/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/],
        // Идентификатор ЖКУ 00АА000000-00.
        zhkuId: [/\d/, /\d/, /[а-яА-ЯЁё]/, /[а-яА-ЯЁё]/, /\d/, /\d/, /\d/, /\d/, /\d/, /\d/, '-', /\d/, /\d/],
        // Идентификатор платёжного документа (ЖКУ) 00АА000000-00-0000.
        zhkuPaymentDocumentId: [
            /\d/,
            /\d/,
            /[а-яА-ЯЁё]/,
            /[а-яА-ЯЁё]/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            /\d/,
            '-',
            /\d/,
            /\d/,
            '-',
            /\d/,
            /\d/,
            /\d/,
            /\d/,
        ],
    },
};
