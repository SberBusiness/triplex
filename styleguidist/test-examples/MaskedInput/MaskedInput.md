```jsx
import React, {useState, useEffect, useRef} from 'react';
import ComponentOptions from '../../common/components/ComponentOptions/ComponentOptions';
import ComponentPreview from '../../common/components/ComponentPreview/ComponentPreview';
import {MaskedInput} from '@sberbusiness/triplex/components/MaskedInput/MaskedInput';

const [value, setValue] = useState('');
const [focused, setFocused] = useState(false);
const [error, setError] = useState(false);
const [disabled, setDisabled] = useState(false);
const ref = useRef();

useEffect(() => {
    if (focused) {
        ref.current.focus();
    }
}, [focused]);

const checkboxOptions = [
    {
        id: 'focused',
        label: 'focused',
        checked: focused,
        onChange: setFocused,
    },
    {
        id: 'error',
        label: 'error',
        checked: error,
        onChange: setError,
    },
    {
        id: 'disabled',
        label: 'disabled',
        checked: disabled,
        onChange: setDisabled,
    },
];

const inputOptions = [
    {
        id: 'value',
        hidden: true,
        value: value,
        onChange: setValue,
    },
];

<>
    <ComponentOptions checkboxOptions={checkboxOptions} inputOptions={inputOptions}/>
    <ComponentPreview>
        <MaskedInput
            value={value}
            mask={MaskedInput.presets.masks.phone}
            onChange={(event) => {
                setValue(event.target.value);
            }}
            forwardRef={ref}
            error={error}
            disabled={disabled}/>
    </ComponentPreview>
</>
```
