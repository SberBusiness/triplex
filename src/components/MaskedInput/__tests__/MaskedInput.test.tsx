import React from 'react';
import renderer from 'react-test-renderer';
import {MaskedInput} from '@sberbusiness/triplex/components/MaskedInput/MaskedInput';

jest.mock('react-text-mask', () => ({
    __esModule: true,
    default: jest.fn(({mask, placeholderChar, ...props}: any) => <input type="text" {...props} />),
    conformToMask: jest.fn((x) => ({conformedValue: x})),
}));

describe('MaskedInput', () => {
    beforeEach(() => {
        allure.feature('MaskedInput');
    });

    it('renders correctly', () => {
        const tree = renderer.create(<MaskedInput value="1234567890" onChange={() => undefined} mask={MaskedInput.presets.masks.phone} />);

        expect(tree.toJSON()).toMatchSnapshot();
        tree.unmount();
    });
});
