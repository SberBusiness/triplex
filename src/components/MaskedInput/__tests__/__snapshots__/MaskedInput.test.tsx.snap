// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskedInput renders correctly 1`] = `
<div
  className="cssClass[maskedInputWrapper]"
>
  <input
    aria-hidden="true"
    className="cssClass[input] cssClass[maskedInputPlaceholder]"
    onChange={[Function]}
    placeholder=""
    readOnly={true}
    tabIndex={-1}
  />
  <input
    className="cssClass[maskedInput]"
    defaultValue=""
    guide={false}
    onChange={[Function]}
    render={[Function]}
    type="text"
    value="1234567890"
  />
</div>
`;
